### PR TITLE
Add opentracing support

### PR DESCRIPTION
**Pull request by mohabusama**
Adding opentracing support

### CODE REVIEW

File: zmon_agent/main.py [hunk 1]
```diff
@@ -176,54 +211,76 @@ def main():
...
-                           'set env variable ZMON_AGENT_INFRASTRUCTURE_ACCOUNT.')
+    logger.info('Initializing opentracing tracer: {}'.format(args.opentracing if args.opentracing else 'noop'))
+    init_opentracing_tracer(args.opentracing)
 
-    region = os.environ.get('ZMON_AGENT_REGION', args.region)
-    entity_service = os.environ.get('ZMON_AGENT_ENTITY_SERVICE_URL', args.entity_service)
-    interval = os.environ.get('ZMON_AGENT_INTERVAL', args.interval)
+    # Give some time for tracer initialization
+    if args.opentracing:
+        time.sleep(2)
```
Comment (lmineiro) [inline zmon_agent/main.py:235]: Nice hack :/

**Comment by lmineiro** [comment]
:+1:

**Comment by vetinari** [comment]
:+1:

File: zmon_agent/main.py (hunk 1 above)
Comment (mohabusama) [reply zmon_agent/main.py:235]: Was needed for some vendor ;)

File: zmon_agent/main.py [hunk 2]
```diff
@@ -97,57 +118,71 @@ def sync(infrastructure_account, region, entity_service, verify, dry_run, interv
 
     while True:
         try:
-            zmon_client = get_clients(entity_service, verify=verify)
+            sync_span = opentracing.tracer.start_span(operation_name='zmon-agent-sync')
+
+            with sync_span:
+                zmon_client = get_clients(entity_service, verify=verify)
```
Comment (vetinari) [inline zmon_agent/main.py]: The span is missing the account, region, team? cluster_id tags

File: zmon_agent/main.py [hunk 3]
```diff
@@ -77,16 +95,19 @@ def add_new_entities(all_current_entities, existing_entities, zmon_client, dry_r
...
                 resp = zmon_client.add_entity(entity)
 
                 resp.raise_for_status()
         except Exception:
+            current_span.set_tag('error', True)
             logger.exception('Failed to add entity!')
+            current_span.log_kv({'exception': traceback.format_exc()})
             error_count += 1
 
+    current_span.log_kv({'new_entities': new_entities})
```
Comment (vetinari) [inline zmon_agent/main.py]: do we really need the full entities here? isn't the id enough?

Comment (mohabusama) [reply zmon_agent/main.py]: Will remove

File: zmon_agent/main.py (hunk 2 above)
Comment (mohabusama) [reply zmon_agent/main.py]: Added to sync_span

File: zmon_agent/discovery/kubernetes/cluster.py [hunk 4]
```diff
@@ -124,16 +134,17 @@ def get_entities(self) -> list:
...
 
         return list(itertools.chain(
             pod_container_entities, node_entities, namespace_entities, service_entities, replicaset_entities,
             daemonset_entities, ingress_entities, statefulset_entities, postgresql_cluster_entities,
             postgresql_cluster_member_entities, postgresql_database_entities))
 
 
+@trace()
 def get_all(kube_client, kube_func, namespace=None) -> list:
     items = []
```
Comment (vetinari) [inline zmon_agent/discovery/kubernetes/cluster.py]: I think it should be (without the `@trace()`):
```python
def get_all(kube_client, kube_func, namespace=None, span=None) -> list:
    if not span:
        span = opentracing.tracer.start_span(operation_name='get_all')
        span.log_kv({'error': 'missing span parameter'})
    with span:
￼        items = []
        ... everything indented one level
```

**Review by vetinari — APPROVED** [review]

**Comment by vetinari** [comment]
:+1:

**Comment by mohabusama** [comment]
👍

**Pull request merged**